rpc/src/module/pool.rs

```diff
@@ -100,20 +100,24 @@ impl PoolRpc for PoolRpcImpl {
             return Err(RPCError::ckb_internal_error(e));
         }
 
+        let broadcast = |tx_hash: packed::Byte32| {
+            // workaround: we are using `PeerIndex(usize::max)` to indicate that tx hash source is itself.
+            let peer_index = PeerIndex::new(usize::max_value());
+            self.sync_shared
+                .state()
+                .tx_hashes()
+                .entry(peer_index)
+                .or_default()
+                .insert(tx_hash);
+        };
+
         match submit_txs.unwrap() {
             Ok(_) => {
-                // workaround: we are using `PeerIndex(usize::max)` to indicate that tx hash source is itself.
-                let peer_index = PeerIndex::new(usize::max_value());
-                let hash = tx.hash();
-                self.sync_shared
-                    .state()
-                    .tx_hashes()
-                    .entry(peer_index)
-                    .or_default()
-                    .insert(hash.clone());
-                Ok(hash.unpack())
+                let tx_hash = tx.hash();
+                broadcast(tx_hash.clone());
+                Ok(tx_hash.unpack())
             }
-            Err(e) => Err(RPCError::from_ckb_error(e)),
+            Err(e) => Err(RPCError::submit_transaction_error(e, broadcast)),
```

Comment:
It's weird to pass `broadcast` to the error conversion function. I would suggest match the returned error and broadcast on the duplication error.